File: go/tree/108_sorted_array_to_bst.go
package tree

// sortedArrayToBST 将有序数组转换为二叉搜索树
func sortedArrayToBST(nums []int) *TreeNode {

	// 中序遍历，以中间节点作为根节点

	return builder(nums, 0, len(nums)-1)
}

func builder(nums []int, start, end int) *TreeNode {

	// 中序遍历，以中间节点作为根节点
	if start > end {
		return nil
	}

	// 奇数根节点为中间点，偶数 为中间节点偏左的
	// not use(start + end) / 2 防止int 越界
	mid := start + (end-start)>>1

	root := &TreeNode{
		Val: nums[mid],
	}

	// BUG this root.Left = builder(nums, 0, mid-1)
	root.Left = builder(nums, start, mid-1)
	root.Right = builder(nums, mid+1, end)
	return root
}

File: rust/src/tree/108_sorted_array_to_bst.rs


// sorted_array_to_bst 中序遍历
pub fn sorted_array_to_bst(nums: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {

    let nums = &nums[..];

    builder(nums, 0,  (nums.len() -1) as i32)
}

fn builder(nums: &[i32], left: i32, right: i32) -> Option<Rc<RefCell<TreeNode>>> {

    if left > right {
        return None
    }

    // BUG 1 + (1-0) >>1
    // let mid = left + (right - left) >> 1;
    //  FIX ABOVE
    let mid  = left + ((right - left)>>1);
    // or below
    // let mid = (right + left) >> 1;


    let root = Some(
        Rc::new(RefCell::new(
            TreeNode{
                val: nums[mid as usize],
                left: None,
                right: None,
            }
        ))
    );

    if let Some(node) = &root {
        // left 用i32 not usize 是为了防止mid = 0 , 0-1 越界
        node.borrow_mut().left = builder(nums, left, mid -1);
        node.borrow_mut().right = builder(nums, mid + 1, right); 
    };

    root
}